src/Avalonia.Controls/ItemsControl.cs

```diff
@@ -226,19 +225,35 @@ protected virtual IItemContainerGenerator CreateItemContainerGenerator()
         /// <param name="e">The details of the containers.</param>
         protected virtual void OnContainersMaterialized(ItemContainerEventArgs e)
         {
-            var toAdd = new List<ILogical>();
-
             foreach (var container in e.Containers)
             {
                 // If the item is its own container, then it will be added to the logical tree when
                 // it was added to the Items collection.
                 if (container.ContainerControl != null && container.ContainerControl != container.Item)
                 {
-                    toAdd.Add(container.ContainerControl);
+                    if (ItemContainerGenerator.ContainerType == null)
+                    {
+                        var containerControl = container.ContainerControl as ContentPresenter;
+
+                        if (containerControl != null)
+                        {
+                            ((ISetLogicalParent)containerControl).SetParent(this);
+                            containerControl.SetValue(TemplatedParentProperty, null);
+
+                            containerControl.UpdateChild();
+
+                            if (containerControl.Child is ILogical)
+                            {
+                                LogicalChildren.Add(containerControl.Child);
```

Comment:
@grokys  Is it more efficient to add them to a list first, then do addrange.

I had at some point during the development needed to add them as I went along for some reason.
might be able to change back if there is a performance hit?